solr/server/build.gradle

```diff
@@ -113,11 +116,12 @@ task assemblePackaging(type: Sync) {
     include "README.md"
   })
 
-  from(configurations.libExt, {
+  // Add in the intersection of serverLib and solrCore
+  from(configurations.libExt + (configurations.serverLib - (configurations.serverLib - configurations.solrCore)), {
```

Comment:
Yeah its not great. I wish there was better syntax for doing an intersection. Maybe my gradle googling just isn't good enough.

But if we subtract (the things in serverLib that aren't in solrCore) for serverLib, that gives us an intersection between serverLib and solrCore. This is the actual de-duplication of jars.

Reply:
Alright so I really must be missing something.

`configurations.libExt + (configurations.serverLib - (configurations.serverLib - configurations.solrCore))`

I think this is simplifying to:

`configurations.libExt - configurations.solrCore`

I feel like `configurations.libExt + (configurations.serverLib - configurations.solrCore)` would be all that is needed? I don't understand the additional `configurations.serverLib` subtraction.

Reply:
This is set math, not arithmetic, so you can't just simplify and remove the parentheses.

1. `(configurations.serverLib - configurations.solrCore)`: This gets the set of `serverLib` jars that are not in `solrCore`
1. `(configurations.serverLib - #1)`: This removes the (set of `serverLib` jars that are not in `solrCore`) from all `serverLib` jars. Therefore we are left with the set of `serverLib` jars that are **not not** (double-negation) in `solrCore` => the set of `serverLib` jars that are in `solrCore`.

Reply:
thanks that helps

Reply:
Your explanation, to me, only seems to prove Kevin's point -- that we're left with libExt + solrCore.  Any way, if Kevin understands then cool with me provided we have a comment to help decipher this riddle.
Curious, did you use println statements or something to debug this?

Reply:
well since I was not 100% sure. Let me try to work out an example:
* `libExt` = `(a b c)`
* `serverLib` = `(c d e)`
* `solrCore` = `(e f g)`

Taking inner parenthesis first: `(configurations.serverLib - configurations.solrCore)`

`(c d e) - (e f g)` = `(c d)`

then next parenthesis: `(configurations.serverLib - result)`

`(c d e) - (c d)` = `(e)`

then final addition: `(configurations.libExt + result)`

`(a b c) + (e)` = `(a b c e)`

So as far as I can tell this matches @HoustonPutman explanation. The set of jars that are in both `solrCore` and `serverLib` added to `libExt`

Reply:
Thanks for the example Kevin. It's because this Set math doesn't follow the associative property, because you can't have negative dependencies 🙂

Reply:
Super clear; thanks!  My error in thinking was that somehow coreLib was a subset of serverLib.  I don't know why I had that in my head :-)